added initial semantic segmentation example

akshaykvnit opened the pull request:
# Before submitting

- [x] Was this discussed/approved via a Github issue? (no need for typos, doc improvements)
- [x] Did you read the [contributor guideline](https://github.com/PyTorchLightning/pytorch-lightning/blob/master/.github/CONTRIBUTING.md)?
- [ ] Did you make sure to update the docs?   
- [ ] Did you write any new necessary tests?  

## What does this PR do?
Fixes #635 (Semantic Segmentation example).

## PR review    
Anyone in the community is free to review the PR once the tests have passed.     
If we didn't discuss your PR in Github issues there's a high chance it will not be merged.

## Some additional info
I tried to add a `test_step` in the `LightningModule` which logs the predicted segmentation output. Since the default logger is TensorBoard, I expected the images to show up when I use `tensorboard logdir=lightning_logs` (at the default save path). But, they don't show up. I couldn't find any example which uses the logger properly. So, if someone can point me to an example, then I'll finish that (right now it's commented out).

Also, I have kind of followed the GAN example as a template, so I have put all of the code in a single  file. If you prefer, then I can separate the `Dataset` class into another file. And I have added this as a `full_example`, so please let me know if you want it in `domain_examples` or something.

Also, I have used the [KITTI dataset](http://www.cvlibs.net/datasets/kitti/) (since it is small and enough for an example) but it is not available directly in `torchvision`. So, it needs to be downloaded separately and the path needs to be given in the `hparams`. Note that the segmentation datasets available in torchvision (like Pascal VOC) are too large for a simple example. 

Also, I couldn't figure out what documentation or tests are required, because none of the other examples have any documentation or tests (that I could find). Please point out if anything is required.

williamFalcon commented:
@akshaykvnit awesome. can you get the tests to pass and update the docs?

Borda commented on pl_examples/full_examples/semantic_segmentation/semseg.py:
```diff
@@ -0,0 +1,151 @@
+
+if __name__ == '__main__':
+    parser = ArgumentParser()
+    parser.add_argument("--root", type = str, help = "path where dataset is stored")
```
I guess lint does not like spaces around =

akshaykvnit replied:
Thanks, I managed to format the things as required by flake8.

Borda commented on pl_examples/full_examples/semantic_segmentation/semseg.py:
```diff
@@ -0,0 +1,151 @@
+    # ------------------------
+    # 3 START TRAINING
+    # ------------------------
+    trainer.fit(model)
```
It would be nice to also add inference usage, like segment an image from dataset...

Borda commented on pl_examples/full_examples/semantic_segmentation/semseg.py:
```diff
@@ -0,0 +1,151 @@
+import pytorch_lightning as pl
+
+class KITTI(Dataset):
+    def __init__(self, root_path, split = 'test', img_size = (1242, 376), transform = None):
```
flake does not like the spaces around =

akshaykvnit replied:
Thanks, I managed to format the things as required by flake8.

Borda commented on pl_examples/full_examples/semantic_segmentation/semseg.py:
```diff
@@ -0,0 +1,151 @@
+        return mask
+    
+    def get_filenames(self, path):
+        files_list = list()
```
pls add docstring

akshaykvnit replied:
Done in next commit

Borda commented on pl_examples/full_examples/semantic_segmentation/semseg.py:
```diff
@@ -0,0 +1,151 @@
+class SegModel(pl.LightningModule):
+    def __init__(self, hparams):
+        super(SegModel, self).__init__()
+        self.hparams = hparams
```
do you need to keep it since you assign all other parameters...

akshaykvnit replied:
Removed in next commit

Borda commented on pl_examples/full_examples/semantic_segmentation/semseg.py:
```diff
@@ -0,0 +1,151 @@
+        loss_val = F.cross_entropy(out['out'], mask, ignore_index = 250)
+        return {'loss' : loss_val}
+    
+#     def test_step(self, batch, batch_nb):
```
why is this commented? if it not needed, pls remove it...

akshaykvnit replied:
This will be needed for the inference usage you mentioned in another comment. But, I haven't been able to get it to work yet.

Borda commented:
@williamFalcon a bit more general thought is it would be better to write examples as ipython notebooks which we can transform to nice pages in Docs even with images as I did [here](https://pyimsegm.readthedocs.io/en/latest/notebooks/segment-2d_slic-fts-clust-gc.html)

akshaykvnit commented:
@Borda @williamFalcon I have made most of the changes suggested by @Borda in the last review (with the exception of inference usage) and got the tests to pass. 

I haven't been able to implement the inference usage properly. That will take some more time. 

> @williamFalcon a bit more general thought is it would be better to write examples as ipython notebooks which we can transform to nice pages in Docs even with images as I did [here](https://pyimsegm.readthedocs.io/en/latest/notebooks/segment-2d_slic-fts-clust-gc.html)

I had initially implemented this in a Jupyter Notebook, but then the usage of `hparams` (using `argparse` and taking input from command line) won't be possible as they have been used in all your examples. So, if we need to convert the examples to iPython notebooks, then I could open a new PR, but do mention what can be done about the `hparams`.

Borda commented:
> I haven't been able to implement the inference usage properly. That will take some more time.

do you want to do it here in this PR or would you prefer to make a follow-up PR?

> I had initially implemented this in a Jupyter Notebook, but then the usage of `hparams` (using `argparse` and taking input from command line) won't be possible as they have been used in all your examples. So, if we need to convert the examples to iPython notebooks, then I could open a new PR, but do mention what can be done about the `hparams`.

In this case, I have in mind to create a kind of lightning ZOO with the models and their particular trainers and then having notebook would be nice as an extra example with hparams can be simple derived without duplicating extra code... just building from prefabricated bricks... @williamFalcon  ?

akshaykvnit commented:
> @akshaykvnit do you want to do it here in this PR or would you prefer to make a follow-up PR?

Yes, I'll make a follow-up PR.

Borda commented on pl_examples/full_examples/semantic_segmentation/semseg.py:
```diff
@@ -0,0 +1,149 @@
+from torch.utils.data import DataLoader, Dataset
+from torchvision.models.segmentation import fcn_resnet50
+
+from PIL import Image
```
pls move PIL to the block above with "3rd-party packages"

akshaykvnit replied:
Done in the latest commit.

Borda commented on pl_examples/full_examples/semantic_segmentation/semseg.py:
```diff
@@ -0,0 +1,149 @@
+
+
+class KITTI(Dataset):
+    def __init__(self, root_path, split='test', img_size=(1242, 376), transform=None):
```
pls add simple docstring what is this class about, in particular, explanation about the void and valid labels
personally I would pass the label indexes as a parameter in init with your listed indexes as default... this may simplify re-usage :]

akshaykvnit replied:
I have added these things in the latest commit.

Borda commented on pl_examples/full_examples/semantic_segmentation/semseg.py:
```diff
@@ -0,0 +1,149 @@
+
+
+class SegModel(pl.LightningModule):
+    def __init__(self, hparams):
```
as examples are becoming part of the Docs, we shall add at least basic documentation with a brief description and reference to resources...

akshaykvnit replied:
I'm not sure what documentation you need here. Because none of the other examples have documentation for the `LightningModule` class.

Borda replied:
I ment someting like:
```
"""Image segmentation model.

This is a basic image segmentation model implemented with Lightning ...

References
- <link to source / publication>
```

akshaykvnit replied:
Done in the latest commit

Borda commented:
> > @akshaykvnit do you want to do it here in this PR or would you prefer to make a follow-up PR?
> 
> Yes, I'll make a follow-up PR.

btw, just curious how did you verify without inference? just see the validation/test curve?

pep8speaks commented:
Hello @akshaykvnit! Thanks for updating this PR. 

* In the file [`pl_examples/full_examples/semantic_segmentation/semseg.py`](https://github.com/PyTorchLightning/pytorch-lightning/blob/aec1d02ff01aecce0ff1f2ad0c0d390436e48045/pl_examples/full_examples/semantic_segmentation/semseg.py):

> [Line 119:101](https://github.com/PyTorchLightning/pytorch-lightning/blob/aec1d02ff01aecce0ff1f2ad0c0d390436e48045/pl_examples/full_examples/semantic_segmentation/semseg.py#L119): [E501](https://duckduckgo.com/?q=pep8%20E501) line too long (112 > 100 characters)
> [Line 122:101](https://github.com/PyTorchLightning/pytorch-lightning/blob/aec1d02ff01aecce0ff1f2ad0c0d390436e48045/pl_examples/full_examples/semantic_segmentation/semseg.py#L122): [E501](https://duckduckgo.com/?q=pep8%20E501) line too long (116 > 100 characters)

akshaykvnit commented:
> > > @akshaykvnit do you want to do it here in this PR or would you prefer to make a follow-up PR?
> > 
> > 
> > Yes, I'll make a follow-up PR.
> 
> btw, just curious how did you verify without inference? just see the validation/test curve?

So, I implemented the reference using normal PyTorch and it works. You can [see the results here](https://github.com/akshaykvnit/pl-sem-seg/blob/master/pl_training.ipynb) in the last block of the notebook. I added the same code after `trainer.fit(model)` and it works properly. Thus, we can be sure that the training works.

However, I figured that you needed inference to work when `trainer.test(model)` is called. I haven't managed to properly implement in this fashion (some problem with the `test_step` and `test_end` functions). However, I shall try to fix that on my own (and open a new PR when done or open an issue if stuck).

Borda commented:
@williamFalcon how about the docs for examples? I would say it shall contain them as it is part of Docs but probably it is not a blocker... :]

williamFalcon commented:
@akshaykvnit awesome tutorial!

williamFalcon commented:
@akshaykvnit 

```
 No module named 'torchvision.models.segmentation'
```

oops, looks like this test fails. Need to fix asap

Borda commented:
Yeah I have noticed that almost all examples are not tested at all... We shall run them with a minimal data....
